### PR TITLE
Vp 1093 personal goal api permissions management casl

### PR DESCRIPTION
**Pull request by avowkind**
## I do solemnly swear that I have:
- [x] Run `npm test` and all the tests pass.
- [x] Run `npm run lint` and there are no warnings.
- [x] Not decreased the overall test coverage?
- [x] Included the Jira ID and description in the PR title

## Proposed Changes? 🤔
Personal Goals API calls are protected by CASL rules
Anon users cannot access personalGoals.
Signed in users can read their own goals but only change status
admin users can read and write personalGoals. - This allows support to fix problems.

### CODE REVIEW

File: server/api/personalGoal/personalGoal.controller.js [hunk 1]
```diff
@@ -24,7 +22,9 @@ const listPersonalGoals = async (req, res) => {
 }
 
 const updatePersonalGoal = async (req, res) => {
-  const status = req.body.status
+  const { body: pg } = req
```
Comment (guttmann) [inline server/api/personalGoal/personalGoal.controller.js]: Is there any reason you don't use `const pg = req.body` here?

File: server/api/personalGoal/__tests__/personalGoal.ability.spec.js [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+  response = await request(server)
+    .get(`/api/personalGoals/${id}`)
+    .send({
+      status: PersonalGoalStatus.ACTIVE
+    })
+
+  t.is(response.statusCode, 403)
+})
+// NORMAL
+test('PersonalGoal API - normal - list self', async t => {
```
Comment (guttmann) [inline server/api/personalGoal/__tests__/personalGoal.ability.spec.js]: What does "normal" mean here? Any authenticated user that isn't an admin?

File: server/api/personalGoal/__tests__/personalGoal.ability.spec.js [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+    .post('/api/personalGoals')
+    .send({
+      person: t.context.people[3]._id,
+      goal: t.context.goals[1]._id
+    })
+
+  t.is(response.statusCode, 403)
+  const id = t.context.personalGoals[1]._id
+  response = await request(server)
+    .get(`/api/personalGoals/${id}`)
```
Comment (guttmann) [inline server/api/personalGoal/__tests__/personalGoal.ability.spec.js]: Is this meant to be `.put`?

File: server/api/personalGoal/__tests__/personalGoal.ability.spec.js [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+
+  t.is(response.statusCode, 403)
+
+  response = await request(server)
+    .get(`/api/personalGoals/${id}`)
+    .send({
+      status: PersonalGoalStatus.ACTIVE
+    })
+
+  t.is(response.statusCode, 403)
```
Comment (guttmann) [inline server/api/personalGoal/__tests__/personalGoal.ability.spec.js]: This looks like a duplication of lines 88-94, is it needed?

File: server/api/personalGoal/personalGoal.controller.js [hunk 5]
```diff
@@ -37,14 +37,18 @@ const updatePersonalGoal = async (req, res) => {
       set.dateCompleted = Date.now()
       break
   }
-  await PersonalGoal.updateOne({ _id: req.body._id }, { $set: set }).exec()
+  try {
+    await PersonalGoal
+      // .accessibleBy(userAbility, Action.UPDATE)
+      .updateOne({ _id: id }, { $set: set }).exec()
```
Comment (guttmann) [inline server/api/personalGoal/personalGoal.controller.js]: As mentioned above, I think this section is currently open to authenticated users updating any other users personal goals

File: server/api/personalGoal/personalGoal.ability.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  const anonAbilities = [{
+    subject: SchemaName,
+    action: Action.CRUD,
+    inverted: true
+  }]
+  const personId = session && session.me && session.me._id ? session.me._id.toString() : undefined
+  const authedAbilities = []
+  const { meid } = query
+
+  if (meid && meid === personId) { // if there is a meid query param we are listing someones pgs
```
[the remark this answers is not in the document]
Comment (avowkind) [reply server/api/personalGoal/personalGoal.ability.js]: I'll try that out.

File: server/api/personalGoal/personalGoal.ability.js [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+  const adminAbilities = [{ subject: SchemaName, action: Action.MANAGE }]
+
+  return {
+    [Role.ANON]: anonAbilities,
+    [Role.VOLUNTEER_PROVIDER]: authedAbilities,
+    [Role.OPPORTUNITY_PROVIDER]: authedAbilities,
+    [Role.ACTIVITY_PROVIDER]: authedAbilities,
+    [Role.ADMIN]: adminAbilities,
+    [Role.ALL]: authedAbilities
```
[the remark this answers is not in the document]
Comment (avowkind) [reply server/api/personalGoal/personalGoal.ability.js]: No its there from copying from the people. I'll remove it

File: server/api/personalGoal/personalGoal.controller.js (hunk 1 above)
Comment (avowkind) [reply server/api/personalGoal/personalGoal.controller.js]: no reason - an earlier version pulled the ability out of the req too.

File: server/api/personalGoal/__tests__/personalGoal.ability.spec.js (hunk 2 above)
Comment (avowkind) [reply server/api/personalGoal/__tests__/personalGoal.ability.spec.js]: yes

File: server/api/personalGoal/__tests__/personalGoal.ability.spec.js (hunk 3 above)
Comment (avowkind) [reply server/api/personalGoal/__tests__/personalGoal.ability.spec.js]: yes - and delete to follow.

File: server/api/personalGoal/personalGoal.ability.js [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+      reason: 'You can only read your own goals'
+    })
+  }
+  // allow update only status fields. put verifies person
+  authedAbilities.push({
+    subject: SchemaName,
+    action: Action.UPDATE,
+    reason: 'You can only change status',
+    conditions: {
+      person: personId
```
[the remark this answers is not in the document]
Comment (avowkind) [reply server/api/personalGoal/personalGoal.ability.js]: added test to read anothers goals.

File: server/api/personalGoal/personalGoal.controller.js (hunk 5 above)
Comment (avowkind) [reply server/api/personalGoal/personalGoal.controller.js]: Resolved by putting the accessibleBy back in.  Using accessibleBy does not cause a 403 Forbidden to be returned. Instead the call returns 200 OK but does not make the changes to the database.  This is ok and the tests can reflect it - but there is a potential for confusing bugs.

Comment (guttmann) [reply server/api/personalGoal/personalGoal.controller.js]: Yeah I've hit that problem elsewhere - I think I handled it something like this:

```
const personalGoal = await PersonalGoal
  .accessibleBy(req.ability, Action.UPDATE)
  .findOne({ _id: id })

if (personalGoal === null) {
  // 403 or 404 depending on preference
}

// update query here
```

The downside being you're doing 2 queries.

Alternatively, it looks like `.updateOne` returns a result object that could be used to check if records were updated by the query

[Mongoose Model.updateOne API doc](https://mongoosejs.com/docs/api/model.html#model_Model.updateOne)

**Pull request merged**